src/queue.rs

```diff
@@ -0,0 +1,819 @@
+        AvailIter {
+            mem,
+            desc_table: GuestAddress(0),
+            avail_ring: GuestAddress(0),
+            next_index: Wrapping(0),
+            last_index: Wrapping(0),
+            queue_size: 0,
+            next_avail: q_next_avail,
+        }
+    }
+}
+
+impl<'a, 'b, M: GuestMemory> Iterator for AvailIter<'a, 'b, M> {
+    type Item = DescriptorChain<'a, M>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.next_index == self.last_index {
+            return None;
+        }
+
+        let offset = (4 + (self.next_index.0 % self.queue_size) * 2) as usize;
+        let avail_addr = match self.mem.checked_offset(self.avail_ring, offset) {
+            Some(a) => a,
+            None => return None,
+        };
+        // This index is checked below in checked_new
+        let desc_index: u16 = match self.mem.read_obj(avail_addr) {
+            Ok(ret) => ret,
+            Err(_) => {
+                // TODO log address
+                error!("Failed to read from memory");
+                return None;
+            }
+        };
+
+        self.next_index += Wrapping(1);
+
+        let ret =
+            DescriptorChain::checked_new(self.mem, self.desc_table, self.queue_size, desc_index);
+        if ret.is_some() {
+            *self.next_avail += Wrapping(1);
+        }
+        ret
+    }
+}
+
+#[derive(Clone)]
+/// A virtio queue's parameters.
+pub struct Queue<'a, M: GuestMemory> {
+    mem: &'a M,
+
+    /// The maximal size in elements offered by the device
+    max_size: u16,
+
+    /// The queue size in elements the driver selected
+    pub size: u16,
+
+    /// Indicates if the queue is finished with configuration
+    pub ready: bool,
+
+    /// Guest physical address of the descriptor table
+    pub desc_table: GuestAddress,
+
+    /// Guest physical address of the available ring
+    pub avail_ring: GuestAddress,
+
+    /// Guest physical address of the used ring
+    pub used_ring: GuestAddress,
+
+    next_avail: Wrapping<u16>,
+    next_used: Wrapping<u16>,
+}
+
+impl<'a, M: GuestMemory> Queue<'a, M> {
+    /// Constructs an empty virtio queue with the given `max_size`.
+    pub fn new(mem: &M, max_size: u16) -> Queue<M> {
+        Queue {
+            mem,
+            max_size,
+            size: max_size,
+            ready: false,
+            desc_table: GuestAddress(0),
+            avail_ring: GuestAddress(0),
+            used_ring: GuestAddress(0),
+            next_avail: Wrapping(0),
+            next_used: Wrapping(0),
+        }
+    }
+
+    /// Gets the virtio queue maximum size.
+    pub fn get_max_size(&self) -> u16 {
+        self.max_size
+    }
+
+    /// Return the actual size of the queue, as the driver may not set up a
+    /// queue as big as the device allows.
+    pub fn actual_size(&self) -> u16 {
+        min(self.size, self.max_size)
+    }
+
+    /// Check if the virtio queue configuration is valid.
+    pub fn is_valid(&self) -> bool {
```

Comment:
What is a typical use of `is_valid`? I am asking because I was thinking whether instead of returning a bool we can change it to `check_valid` and return a Result. With this approach the caller of `check_valid` can decide what errors are irrecoverable and for which to retry. Not sure if it makes sense though. Let me know.

Reply:
`is_valid()` is typically called when the transport layer decides it should activate the underlying device, i.e. when the driver told us it's ready to take some IO and we're about to activate the device.

So `is_valid()` assumes a queue has been initialized and fully configured. So it tests the whole configuration validity and anything that's not properly configured would prevent the device or the driver to correctly use the queue. That's why I think a boolean makes sense here.